File: ts-sdk/src/protocol/pagination.ts
/**
 * S18 — Cursor-based pagination (§12).
 *
 * Defines the shapes and utilities for the uniform page-by-page exchange:
 *   - `PaginatedRequestParams`: base request params carrying an optional `cursor`
 *   - `PaginatedResult`: base result shape carrying an optional `nextCursor`
 *   - Cursor-presence and end-of-results predicates
 *   - Cache-key helper enforcing per-cursor cache isolation
 *   - Invalid-cursor error builder (`-32602`)
 *
 * The `Cursor` type itself is the canonical opaque-string type defined in S04
 * (§3.7, Appendix E). S18 governs how it is threaded through list exchanges.
 *
 * Paginated methods: `tools/list`, `resources/list`, `resources/templates/list`,
 * `prompts/list` — each extends these base shapes with its list-specific payload.
 */

import { z } from 'zod';
import { CursorSchema } from '../jsonrpc/payload.js';
import { ResultTypeSchema } from '../jsonrpc/payload.js';

export { CursorSchema, type Cursor } from '../jsonrpc/payload.js';

// ─── PaginatedRequestParams ───────────────────────────────────────────────────

/**
 * Base parameter shape for any paginated list request. (§12, §6 / S18)
 *
 * `cursor` (OPTIONAL): when present the server returns results positioned after
 * this cursor (R-12.2-a). When absent the server returns the first page
 * (R-12.2-b). The empty string `""` is a valid present cursor (R-12.1-a).
 *
 * `_meta` (OPTIONAL): metadata; the per-request `_meta` requirement (§4.3 / S05)
 * applies when this shape is used in a client request — see `RequestParamsSchema`
 * in S04.
 *
 * `.passthrough()` allows method-specific params to survive parse.
 */
export const PaginatedRequestParamsSchema = z
  .object({
    /** OPTIONAL per-request metadata (see §4 / S05 for required keys on client requests). */
    _meta: z.record(z.unknown()).optional(),
    /** OPTIONAL opaque pagination token; absent means first page. (R-12.2-a, R-12.2-b) */
    cursor: CursorSchema.optional(),
  })
  .passthrough();

export type PaginatedRequestParams = z.infer<typeof PaginatedRequestParamsSchema>;

// ─── PaginatedResult ──────────────────────────────────────────────────────────

/**
 * Base result shape for any paginated list result. (§12 / S18)
 *
 * `nextCursor` (OPTIONAL): when present, more results MAY follow (R-12.2-c)
 * and the client uses this exact value as `cursor` on the next request.
 * When absent, this is the final page (R-12.2-d, R-12.3-c).
 *
 * The empty string `""` is a valid `nextCursor` (R-12.3-d) and MUST be sent
 * back as `cursor` to continue — it is NOT an end-of-results signal.
 *
 * Method-specific list members (e.g. `tools`, `resources`) are preserved via
 * `.passthrough()`.
 */
export const PaginatedResultSchema = z
  .object({
    /** REQUIRED discriminator inherited from the base Result shape. (§3.6 / S04) */
    resultType: ResultTypeSchema,
    /** OPTIONAL metadata. */
    _meta: z.record(z.unknown()).optional(),
    /** OPTIONAL: opaque token for the next page; absent means last page. (R-12.2-c, R-12.2-d) */
    nextCursor: CursorSchema.optional(),
  })
  .passthrough();

export type PaginatedResult = z.infer<typeof PaginatedResultSchema>;

// ─── Cursor predicates ────────────────────────────────────────────────────────

/**
 * Returns `true` when `nextCursor` is present in the result, indicating that
 * more results MAY be available and the client should request the next page.
 * (R-12.2-c, R-12.3-b)
 *
 * Note: the empty string `""` is a PRESENT cursor — it is NOT treated as
 * absence. (R-12.1-a, R-12.3-d)
 */
export function hasNextCursor(result: { nextCursor?: string }): boolean {
  return result.nextCursor !== undefined;
}

/**
 * Returns `true` when this is the final page — `nextCursor` is absent.
 * (R-12.2-d, R-12.3-c)
 */
export function isLastPage(result: { nextCursor?: string }): boolean {
  return !hasNextCursor(result);
}

/**
 * Returns `true` when `cursor` is a present value (including the empty string).
 *
 * A client MUST treat a present `nextCursor` — even `""` — as a cursor to
 * echo back on the next request. Only `undefined` signals "no cursor". (R-12.1-a)
 */
export function isCursorPresent(cursor: string | undefined): boolean {
  return cursor !== undefined;
}

// ─── Invalid-cursor error ─────────────────────────────────────────────────────

/** Error code for an invalid / unrecognized cursor. (R-12.4-c) */
export const INVALID_CURSOR_CODE = -32602 as const;

/**
 * Builds the JSON-RPC error payload for an invalid-cursor rejection.
 * (R-12.4-c, R-12.4-d)
 *
 * Servers SHOULD return this code when a client supplies a cursor that was
 * not issued by this server, is not recognized, or is otherwise malformed.
 */
export function buildInvalidCursorError(message?: string): {
  code: typeof INVALID_CURSOR_CODE;
  message: string;
} {
  return {
    code: INVALID_CURSOR_CODE,
    message: message ?? 'Invalid params: unrecognized cursor',
  };
}

// ─── Cache-key helper ─────────────────────────────────────────────────────────

/**
 * Produces a cache entry key for a paginated request.
 *
 * Each page of a paginated result is an independent cacheable response keyed
 * by the request that produced it (including the `cursor` value). A cached page
 * for one `cursor` value MUST NOT be served as the response for a request bearing
 * a different `cursor` value (including the first-page request, which omits
 * `cursor`). (R-12.5-a, R-13.5-i)
 *
 * Callers may use this to implement per-page cache entries or to verify that
 * two requests do NOT share a cache entry.
 */
export function paginationCacheKey(method: string, cursor: string | undefined): string {
  return cursor === undefined
    ? `${method}::page:first`
    : `${method}::page:cursor:${cursor}`;
}

// ─── Reference paginator ─────────────────────────────────────────────────────

/**
 * Discriminated result from `OffsetPaginator.getPage`.
 *
 * On success (`ok: true`): the page items and an optional next cursor.
 * On failure (`ok: false`): a structured error from `buildInvalidCursorError` —
 *   the paginator never throws on unrecognized cursors. (RC-3, RC-4)
 */
export type PaginatorPageResult<T> =
  | { ok: true; items: ReadonlyArray<T>; nextCursor: string | undefined }
  | { ok: false; error: ReturnType<typeof buildInvalidCursorError> };

/**
 * Reference implementation of a cursor-based paginator over an in-memory array.
 *
 * Cursors are deterministic decimal offset strings — the same position always
 * yields the same cursor token (RC-2: stability). An unrecognized or malformed
 * cursor is returned as a structured error rather than thrown (RC-3, RC-4).
 */
export class OffsetPaginator<T> {
  /** Items returned per page. */
  readonly pageSize: number;

  constructor(
    private readonly items: ReadonlyArray<T>,
    pageSize = 20,
  ) {
    if (!Number.isInteger(pageSize) || pageSize < 1) {
      throw new RangeError('pageSize must be a positive integer');
    }
    this.pageSize = pageSize;
  }

  /**
   * Returns a page of items for the given cursor.
   *
   * - Absent cursor (`undefined`) → first page. (R-12.2-b)
   * - Present cursor → page starting at the encoded offset. (R-12.2-a)
   * - Unrecognized cursor → `{ ok: false, error }` — never throws. (RC-3, RC-4)
   */
  getPage(cursor: string | undefined): PaginatorPageResult<T> {
    const offset = cursor === undefined ? 0 : this.decodeCursor(cursor);
    if (offset === null) {
      return { ok: false, error: buildInvalidCursorError() };
    }

    const page = this.items.slice(offset, offset + this.pageSize);
    const nextOffset = offset + this.pageSize;
    const nextCursor =
      nextOffset < this.items.length ? this.encodeCursor(nextOffset) : undefined;

    return { ok: true, items: page, nextCursor };
  }

  /** Encodes an offset as a deterministic decimal cursor string. (RC-2) */
  private encodeCursor(offset: number): string {
    return String(offset);
  }

  /** Decodes a cursor string; returns `null` for any unrecognized token. */
  private decodeCursor(cursor: string): number | null {
    if (!/^\d+$/.test(cursor)) return null;
    const n = Number(cursor);
    if (n < 0 || n > this.items.length || !Number.isInteger(n)) return null;
    return n;
  }
}

// ─── Paginated methods registry ───────────────────────────────────────────────

/**
 * The set of method names whose results carry `PaginatedResult` shapes. (§12)
 *
 * All four methods support optional `cursor` on the request and optional
 * `nextCursor` on the result. Their list-specific payloads are defined in the
 * respective feature stories (S24, S26, S28).
 */
export const PAGINATED_METHODS = new Set([
  'tools/list',
  'resources/list',
  'resources/templates/list',
  'prompts/list',
] as const);

/** Returns `true` when `method` is one of the paginated list methods. */
export function isPaginatedMethod(method: string): boolean {
  return PAGINATED_METHODS.has(method as never);
}

File: py-sdk/src/mcp/protocol/pagination.py
"""Cursor-based pagination (§12).

Utilities for the uniform page-by-page exchange: cursor-presence / end-of-results
predicates, the per-cursor cache-key helper, the invalid-cursor (-32602) error, and a
reference offset paginator. The ``Cursor`` type is the opaque string from §3.7.

Paginated methods: ``tools/list``, ``resources/list``, ``resources/templates/list``,
``prompts/list``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any

from pydantic import Field

from mcp._model import McpModel, validates
from mcp.jsonrpc.payload import is_cursor

# Error code for an invalid / unrecognized cursor. (R-12.4-c) — same value as -32602.
INVALID_CURSOR_CODE = -32602


# ─── Cursor (re-export) ───────────────────────────────────────────────────────
# The ``Cursor`` opaque-string type is canonical in S04 (§3.7); :func:`is_cursor`
# is its validator. Pagination re-exports it so callers can validate a cursor token
# without reaching into :mod:`mcp.jsonrpc.payload` (parity with the TS
# ``CursorSchema`` re-export from ``pagination.ts``).
__all__ = [
  "INVALID_CURSOR_CODE",
  "PAGINATED_METHODS",
  "PaginatedRequestParams",
  "PaginatedResult",
  "OffsetPaginator",
  "PaginatorPage",
  "build_invalid_cursor_error",
  "has_next_cursor",
  "is_cursor",
  "is_cursor_present",
  "is_last_page",
  "is_paginated_method",
  "is_valid_paginated_request_params",
  "is_valid_paginated_result",
  "pagination_cache_key",
]


# ─── PaginatedRequestParams / PaginatedResult structural validators ───────────

class PaginatedRequestParams(McpModel):
  """The base parameter shape for any paginated list request (§12 / S18) — the Python
  analogue of the TS ``PaginatedRequestParamsSchema``.

  ``cursor`` is OPTIONAL; when present it MUST be an opaque string (``""`` is a valid
  PRESENT cursor, R-12.1-a). Method-specific members are tolerated.

  Validates only the S18-owned base shape: the per-request ``_meta`` REQUIRED-key
  constraint (§4.3 / S05) is enforced by ``is_valid_request_params`` when this rides a
  client request, so a first-page request (``{}``) parses as valid here. (R-12.2-a/-b)
  """

  cursor: str | None = None
  meta: dict[str, Any] | None = Field(default=None, alias="_meta")


class PaginatedResult(McpModel):
  """The base result shape paginated list results extend (§12 / S18) — the Python analogue
  of the TS ``PaginatedResultSchema``.

  REQUIRED string ``resultType`` (the §3.6 discriminator); OPTIONAL opaque-string
  ``nextCursor`` (``""`` signals continuation, NOT end — R-12.3-d). Method-specific list
  members are tolerated. (R-12.2-c, R-12.2-d)
  """

  result_type: str
  next_cursor: str | None = None
  meta: dict[str, Any] | None = Field(default=None, alias="_meta")


def is_valid_paginated_request_params(value: object) -> bool:
  """Return ``True`` for valid base paginated-request ``params``. (§12, §6 / S18)"""
  return validates(PaginatedRequestParams, value)


def is_valid_paginated_result(value: object) -> bool:
  """Return ``True`` for a valid base paginated ``result`` shape. (§12 / S18)"""
  return validates(PaginatedResult, value)


def has_next_cursor(result: dict) -> bool:
  """Return ``True`` when ``nextCursor`` is present (even ``""``), i.e. more may follow.

  The empty string is a PRESENT cursor — only absence signals the last page.
  (R-12.2-c, R-12.3-d)
  """
  return "nextCursor" in result


def is_last_page(result: dict) -> bool:
  """Return ``True`` when this is the final page — ``nextCursor`` is absent. (R-12.2-d)"""
  return not has_next_cursor(result)


def is_cursor_present(cursor: str | None) -> bool:
  """Return ``True`` when ``cursor`` is a present value (including ``""``). Only ``None``
  signals "no cursor". (R-12.1-a)
  """
  return cursor is not None


def build_invalid_cursor_error(message: str | None = None) -> dict:
  """Build the JSON-RPC error payload for an invalid-cursor rejection. (R-12.4-c/-d)"""
  return {"code": INVALID_CURSOR_CODE, "message": message or "Invalid params: unrecognized cursor"}


def pagination_cache_key(method: str, cursor: str | None) -> str:
  """Produce a per-page cache key. A page cached for one ``cursor`` MUST NOT be served
  for a request bearing a different ``cursor`` (including the first-page request which
  omits it). (R-12.5-a, R-13.5-i)
  """
  return f"{method}::page:first" if cursor is None else f"{method}::page:cursor:{cursor}"


@dataclass(frozen=True)
class PaginatorPage:
  """Outcome of :meth:`OffsetPaginator.get_page`.

  On success ``ok`` is ``True`` with ``items`` + ``next_cursor``; on an unrecognized
  cursor ``ok`` is ``False`` with a structured ``error`` (never raised).
  """

  ok: bool
  items: list = field(default_factory=list)
  next_cursor: str | None = None
  error: dict | None = None


class OffsetPaginator:
  """Reference cursor paginator over an in-memory list. Cursors are deterministic decimal
  offset strings (stable); an unrecognized cursor yields a structured error, never raises.
  """

  def __init__(self, items: list, page_size: int = 20) -> None:
    if not isinstance(page_size, int) or page_size < 1:
      raise ValueError("page_size must be a positive integer")
    self._items = list(items)
    self.page_size = page_size

  def get_page(self, cursor: str | None) -> PaginatorPage:
    """Return a page for ``cursor`` (absent → first page); unrecognized → error result."""
    offset = 0 if cursor is None else self._decode_cursor(cursor)
    if offset is None:
      return PaginatorPage(False, error=build_invalid_cursor_error())
    page = self._items[offset : offset + self.page_size]
    next_offset = offset + self.page_size
    next_cursor = str(next_offset) if next_offset < len(self._items) else None
    return PaginatorPage(True, items=list(page), next_cursor=next_cursor)

  def _decode_cursor(self, cursor: str) -> int | None:
    if not re.match(r"^\d+$", cursor):
      return None
    n = int(cursor)
    return n if 0 <= n <= len(self._items) else None


#: Method names whose results carry paginated shapes. (§12)
PAGINATED_METHODS = frozenset({"tools/list", "resources/list", "resources/templates/list", "prompts/list"})


def is_paginated_method(method: str) -> bool:
  """Return ``True`` when ``method`` is one of the paginated list methods."""
  return method in PAGINATED_METHODS
